lib/Doctrine/DBAL/Connections/PrimaryReplicaConnection.php

```diff
@@ -0,0 +1,361 @@
+            throw new InvalidArgumentException('You have to configure at least one replica.');
+        }
+
+        $params['primary']['driver'] = $params['driver'];
+        foreach ($params['replica'] as $replicaKey => $replica) {
+            $params['replica'][$replicaKey]['driver'] = $params['driver'];
+        }
+
+
+        parent::__construct($params, $driver, $config, $eventManager);
+    }
+
+    private function switchToPrimary(): bool
+    {
+        $forcePrimaryAsReplica = $this->getTransactionNestingLevel() > 0;
+
+        if (isset($this->primary)) {
+            $this->_conn = $this->primary;
+
+            if ($forcePrimaryAsReplica) {
+                $this->replica = $this->_conn;
+            }
+
+            return false;
+        }
+
+        $this->primary = $this->_conn = $this->connectTo('primary');
+
+        // Set replica connection to primary to avoid invalid reads
+        $this->replica = $this->primary;
+
+        if ($this->_eventManager->hasListeners(Events::postConnect)) {
+            $eventArgs = new ConnectionEventArgs($this);
+            $this->_eventManager->dispatchEvent(Events::postConnect, $eventArgs);
+        }
+
+        return true;
+    }
+
+    public function connect(): bool
+    {
+        $connectionName = 'replica';
+
+        if ($this->_conn !== null) {
+            return false;
+        }
+
+        $forcePrimaryAsReplica = false;
+
+        if ($this->getTransactionNestingLevel() > 0) {
+            $connectionName        = 'primary';
+            $forcePrimaryAsReplica = true;
+        }
+
+        if (isset($this->$connectionName)) {
+            $this->_conn = $this->$connectionName;
+
+            if ($forcePrimaryAsReplica) {
+                $this->replica = $this->_conn;
+            }
+
+            return false;
+        }
+
+        if ($connectionName === 'primary') {
+            $this->primary = $this->_conn = $this->connectTo($connectionName);
+
+            // Set replica connection to primary to avoid invalid reads
+            $this->replica = $this->primary;
+        } else {
+            $this->replica = $this->_conn = $this->connectTo($connectionName);
+        }
+
+        if ($this->_eventManager->hasListeners(Events::postConnect)) {
+            $eventArgs = new ConnectionEventArgs($this);
+            $this->_eventManager->dispatchEvent(Events::postConnect, $eventArgs);
+        }
+
+        return true;
+    }
+
+    /**
+     * Connects to a specific connection.
+     */
+    private function connectTo(string $connectionName): DriverConnection
+    {
+        $params = $this->getParams();
+
+        $driverOptions = $params['driverOptions'] ?? [];
+
+        $connectionParams = $this->chooseConnectionConfiguration($connectionName, $params);
+
+        $user     = $connectionParams['user'] ?? null;
+        $password = $connectionParams['password'] ?? null;
+
+        return $this->_driver->connect($connectionParams, $user, $password, $driverOptions);
```

Comment:
And this is another place where the abstraction leaks. Shouldn't the parent class implement the logic of converting the DBAL parameters to a `Driver::connect()` call?

https://github.com/doctrine/dbal/blob/b0ca9434e5429079f02f661f5aefea2021cd1c75/src/Connection.php#L350-L354